book/src/rfc/rfc-0-generic-parameters.md

```diff
@@ -0,0 +1,359 @@
+```rust
+use std::bits;
+
+const num_bits = 8;
+
+fn main() {
+    let val1 = 101;
+    // `num_bits` will be assigned to the generic parameter `LEN` in the return type
+    // then the type of `bits` will be monomorphized to [Bool; 8]
+    let bits = bits::to_bits(num_bits, val); 
+    // the value of `LEN` can be determined from the size of `bits` during monomorphization
+    // so the builtin function knows how many bits to convert
+    let val2 = bits::from_bits(bits); 
+    assert_eq(val1, val2);
+}
+```
+
+
+The values for the generic parameters will be passed to the function via the `generics` argument:
+
+```rust
+fn to_bits<B: Backend>(
+    compiler: &mut CircuitWriter<B>,
+    generics: &GenericParameters,
+    vars: &[VarInfo<B::Field, B::Var>],
+    span: Span,
+) -> Result<Option<Var<B::Field, B::Var>>> {
+    ...
+
+    // retrieve the generic values from the `generics` argument
+    let bitlen = generics.get("LEN") as usize;
+
+    ...
+}
+```
+
+Both the return type and returned vars can be checked outside of the builtin functions. The return type can be checked automatically in the same way as the native functions, the types of which are propagated and converged at certain point, at which error will be thrown if the types are not matched.
+
+The return vars can be checked by relying on the types. Each concrete type has a fixed number of vars. With the resolved return type of the builtin function, we can check if the size is matched. Additionally, we can check the values recursively with the type structure, but it might only limited to checking the boolean type which got obvious bound 0 or 1. So automatically checking if the actual return from a builtin is an area to be improved in the future.
+
+
+## monomorphization
+The resolving of the generic values can be done by observing the arguments passed to the function. Then it stores the resolved values in the relevant contexts for the following compiler pipeline to do type checking and circuit synthesizing. We call the process of resolving the generic values and type checking _Monomorphization_.
+
+The current pipeline of compiling noname code is:
+1. Parse the code into AST
+2. Convert the AST into NAST with naming resolution
+3. Convert the NAST into TAST with type metadata collection and type checking
+4. Circuit synthesizing TAST into an constraint system
+
+With generic parameters, the current TAST phase can't handle the type checking anymore, because the generic parameters are unknown. For example, it can't type check the array with symbolic size without resolving the values for the generic parameters.
+
+To extend the type checking to support generic parameters, we can add a MAST phase (Monomorphized AST), right after the TAST phase, to resolve the values of the generic parameters. The circuit synthesizer will rely on the MAST instead of the TAST to compile a circuit.
+
+The new pipeline will be: AST Parser -> NAST -> TAST -> MAST -> Circuit synthesizing
```

Comment:
Maybe we could use the diagram that I made

Reply:
added